staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/optionalwaitgroup.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairqueuing
+
+import (
+	"sync"
+)
+
+// OptionalWaitGroup is a WaitGroup or nothing.  Use it to track when
+// goroutines start or stop or begin or end waiting if the pointer is
+// not nil.  This abstraction is intended to make it easier to write a
+// package that can be unit tested by a function that advances a fake
+// clock once all the available work is done.  Real clients set the
+// pointer to nil because they do not need this tracking.
+type OptionalWaitGroup struct {
```

Comment:
it doesn't need to be an exported type i suppose

Reply:
It is used in other packages, so it needs to be exported.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/waitgroup/waitgroup.go

consider move it to another exisiting package

Reply:
OK.  What package would you suggest?

Reply:
> https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/waitgroup/waitgroup.go

under staging, apimachinery repo, there's already a wait-group package

Reply:
I expect push-back from putting it there, just like I got push-back to putting my new kind of clock where the regular clock is.